atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/cache/ValueStoreImpl.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.keyvalue.api.cache;
+
+import com.palantir.atlasdb.keyvalue.api.AtlasLockDescriptorUtils;
+import com.palantir.atlasdb.keyvalue.api.CellReference;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.lock.LockDescriptor;
+import com.palantir.lock.watch.LockEvent;
+import com.palantir.lock.watch.LockWatchCreatedEvent;
+import com.palantir.lock.watch.LockWatchEvent;
+import com.palantir.lock.watch.LockWatchReferences.LockWatchReference;
+import com.palantir.lock.watch.LockWatchReferencesVisitor;
+import com.palantir.lock.watch.UnlockEvent;
+import com.palantir.logsafe.Preconditions;
+import com.palantir.logsafe.UnsafeArg;
+import io.vavr.collection.HashMap;
+import io.vavr.collection.HashSet;
+import io.vavr.collection.Map;
+import io.vavr.collection.Set;
+import java.util.stream.Stream;
+
+public final class ValueStoreImpl implements ValueStore {
+    // TODO(jshah): implement cache eviction based on cache size
+    private final StructureHolder<Map<CellReference, CacheEntry>> values;
+    private final StructureHolder<Set<TableReference>> watchedTables;
+    private final LockWatchVisitor visitor = new LockWatchVisitor();
+
+    public ValueStoreImpl() {
+        values = StructureHolder.create(HashMap::empty);
```

Comment:
I actually have a counter-proposal:

```java
private final StructureHolder<io.vavr.collection.Map<CellReference, CacheEntry>> values;
```

Then we don't need to express that for the `HashMap`, since it must be an implementation of vavr's Map.